File: tmp/fluent_python/Complex.py
class Complex:
    r = 5.0
    i = 5.0
 
    def print(self):
        print(f'Complex : {self.r} + {self.i}i')
 
    def change(self, r, i):
        self.r = r
        self.i = i

c1 = Complex()
c2 = Complex()
 
c1.print()
c2.print()

# c1 객체는 같은 이름의 클래스 변수, 객체 변수가 있기 때문에 
# print 메소드에서는 이제 객체 변수에 우선적으로 접근
c1.r = 10.0
c1.i = 7.0
 
c1.print()
c2.print()

File: c++/test/complex.cpp
#include <iostream>

class Complex {
 private:
  double real, img;

 public:
  Complex(double real, double img) : real(real), img(img) {}
  Complex(const Complex& c) { real = c.real, img = c.img; }

  Complex operator+(const Complex& c) const;
  Complex operator-(const Complex& c) const;
  Complex operator*(const Complex& c) const;
  Complex operator/(const Complex& c) const;

  Complex& operator=(const Complex& c);
  void println() { std::cout << "( " << real << " , " << img << " ) " << std::endl; }
};

Complex Complex::operator+(const Complex& c) const {
  Complex temp(real + c.real, img + c.img);
  return temp;
}
Complex Complex::operator-(const Complex& c) const {
  Complex temp(real - c.real, img - c.img);
  return temp;
}
Complex Complex::operator*(const Complex& c) const {
  Complex temp(real * c.real - img * c.img, real * c.img + img * c.real);
  return temp;
}
Complex Complex::operator/(const Complex& c) const {
  Complex temp(
    (real * c.real + img * c.img) / (c.real * c.real + c.img * c.img),
    (img * c.real - real * c.img) / (c.real * c.real + c.img * c.img));
  return temp;
}
// 디폴트 대입 연산자(default assignment operator)를 지원하기에 아래와 같이 정의하지 않아도 동작함
Complex& Complex::operator=(const Complex& c) {
  real = c.real;
  img = c.img;
  return *this;
}

int main() {
  Complex a(1.0, 2.0);
  Complex b(3.0, -2.0);
  Complex c(0.0, 0.0);
  c = a * b + a / b + a + b;
  c.println();
}